File: template_tries.py
class BTrieNode:
    def __init__(self):
        self.children = [None, None]  # Index 0 -> bit 0, Index 1 -> bit 1
        self.is_end_of_number = False

class BinaryTrie:
    def __init__(self):
        self.root = BTrieNode()

    def insert(self, number):
        node = self.root
        for i in reversed(range(32)):  # assuming 32-bit integers
            bit = (number >> i) & 1
            if not node.children[bit]:
                node.children[bit] = BTrieNode()
            node = node.children[bit]
        node.is_end_of_number = True

    def search(self, number):
        node = self.root
        for i in reversed(range(32)):
            bit = (number >> i) & 1
            if not node.children[bit]:
                return False
            node = node.children[bit]
        return node.is_end_of_number

    def find_max_xor(self, number):
        """
        Finds the number in the trie that gives the maximum XOR with the given number.
        """
        node = self.root
        if not node.children[0] and not node.children[1]:
            return None  # Trie is empty

        max_xor = 0
        for i in reversed(range(32)):
            bit = (number >> i) & 1
            toggled_bit = 1 - bit
            if node.children[toggled_bit]:
                max_xor |= (1 << i)
                node = node.children[toggled_bit]
            else:
                node = node.children[bit]
        return max_xor

File: cpp-templates/template_tries.cpp
#include <bits/stdc++.h>
using namespace std;

struct BinaryTrie {
    struct Node {
        int child[2];
        bool is_end;
        Node() : child{-1, -1}, is_end(false) {}
    };

    vector<Node> trie;
    int bit_width;

    BinaryTrie(int bit_width_ = 32) : bit_width(bit_width_) {
        trie.emplace_back();
    }

    void insert(int number) {
        int node = 0;
        for (int i = bit_width - 1; i >= 0; i--) {
            int bit = (number >> i) & 1;
            if (trie[node].child[bit] == -1) {
                trie[node].child[bit] = (int)trie.size();
                trie.emplace_back();
            }
            node = trie[node].child[bit];
        }
        trie[node].is_end = true;
    }

    bool search(int number) const {
        int node = 0;
        for (int i = bit_width - 1; i >= 0; i--) {
            int bit = (number >> i) & 1;
            if (trie[node].child[bit] == -1) return false;
            node = trie[node].child[bit];
        }
        return trie[node].is_end;
    }

    optional<int> find_max_xor(int number) const {
        if (trie[0].child[0] == -1 && trie[0].child[1] == -1) return nullopt;
        int node = 0, max_xor = 0;
        for (int i = bit_width - 1; i >= 0; i--) {
            int bit = (number >> i) & 1;
            int want = bit ^ 1;
            if (trie[node].child[want] != -1) {
                max_xor |= (1 << i);
                node = trie[node].child[want];
            } else {
                node = trie[node].child[bit];
            }
        }
        return max_xor;
    }
};
